MF-164 - Extract users service

anovakovic01 opened the pull request:
**Summary:**
- Add users service
- Add HTTP API for users service
- Add gRPC API for users service
- Add logging
- Add metrics
- Add README and swagger files
- Add tests for service implementation
- Add tests for API layer

anovakovic01 committed:
Add users service interfaces and domain logic

Add interfaces for users service, hasher, identity provider and
repository. Add users service domain logic.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add hasher implementation to users service

Add hasher implementation utilising bcrypt.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add identity provider implementation to users service

Add identity provider implementation utilising JWT.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add users repository implementation

Add users repository implementation utilising postgres.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add HTTP endpoints for users service

Add HTTP endpoints for users service with request decoding and
response encoding.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add gRPC API for users service

Split API implementation on gRPC and HTTP. Add gRPC service definition.
Serve gRPC server in main.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add logger to transport layer

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

dborovcanin committed:
MF-136 - Setup staging and test environments (#226)

anovakovic01 committed:
Add tests for users service

Add tests for users service transport layer. Add tests for users
service.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Rename users gRPC service factory function

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add logging and metrics to users service

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add README and HTTP API documentation

Add README file. Add swagger docs.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd reviewed (commented):
For now it looks good, I will take some more time to look it more thoroughly.

One remark though, you should (re)move manager service code that's related to the users service.

drasko reviewed (commented):
Looks good, no remarks from my side.

I note however that Identity is moved to `gRPC`-only, no HTTP counterpart. This will break all other services (adapters), so they must be changed simultaneously with this PR.

anovakovic01 committed:
Fix run and compile command in README file

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Rename gRPC identity method to identify and update service spec

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd commented on users/README.md:
````diff
@@ -0,0 +1,78 @@
+MF_USERS_DB_HOST=[Database host address] MF_USERS_DB_PORT=[Database host port] MF_USERS_DB_USER=[Database user] MF_USERS_DB_PASS=[Database password] MF_USERS_DB=[Name of the database used by the service] MF_USERS_HTTP_PORT=[Service HTTP port] MF_USERS_GRPC_PORT=[Service gRPC port] MF_USERS_SECRET=[String used for signing tokens] $GOBIN/mainflux-users
+```
+
+## Usage
````
Eventually, we should deprecate these "single service API docs", and replace them with one, unified HTTP API doc.

mijicd committed:
Update protobuf to 1.1.0 (#251)

anovakovic01 committed:
NOISSUE - Fix run and compile commands in README files (#249)

anovakovic01 committed:
Add gRPC calls from manager to users service

Add gRPC client calls from manager to users service. Add new
environment variable for users service address.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Fix tests for manager's API tests

Add users client mock implementation. Fix endpoint tests and request
tests.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Remove token response from manager API

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Fix manager service tests

Remove users from manager tests.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Fix client and channel repository tests

Remove users from client and channel repository tests.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd commented on cmd/users/main.go:
```diff
@@ -0,0 +1,129 @@
+			logger.Error(fmt.Sprintf("Failed to listen on port %s: %s", cfg.GRPCPort, err))
+		}
+		server := grpc.NewServer()
+		service := grpcapi.NewService(svc)
```
Don't name it, it has only one usage. It is a bit confusing to have `server`, `service` and `svc`.

mijicd commented on cmd/users/main.go:
```diff
@@ -0,0 +1,129 @@
+	}()
+
+	// Start gRPC server
+	go func() {
```
Since this is quite big now, i propose that you extract these snippets to non-exported functions and use `main()` just to invoke them.

mijicd reviewed (changes requested):
Looks good. Fix these minor remarks and we can merge it. Side note: let's drop the "usage" section from individual services, and let's unify all of the individual API docs into the top-level API specification. There's not much sense into fiddling with individual service.

anovakovic01 committed:
Remove error from gRPC response payload

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Remove error from gRPC response on client side

Remove checking of error from gRPC payload in manager.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Remove error from gRPC response in mock implementation

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Replace Id with ID in variable names

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Update manager to use gRPC getter method

Use gRPC getter method instead of plain value, because this way we
avoid nil pointer exceptions.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add gokit gRPC server to users service

Add endpoint, request, response and transport layer similar to http.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add grpc necessary dependencies

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add client implementation to users gRPC service

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Move HTTP and gRPC serving to separate functions

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd commented on cmd/users/main.go:
```diff
@@ -0,0 +1,130 @@
+
+	logger := log.New(os.Stdout)
+
+	db, err := postgres.Connect(cfg.DBHost, cfg.DBPort, cfg.DBName, cfg.DBUser, cfg.DBPass)
```
Please complete extracting pieces from `main`.

mijicd commented on cmd/users/main.go:
```diff
@@ -0,0 +1,130 @@
+	if err != nil {
+		logger.Error(fmt.Sprintf("Failed to listen on port %s: %s", port, err))
+	}
+	baseServer := grpc.NewServer()
```
Please keep the name `server`.

mijicd commented on manager/postgres/channels_test.go:
```diff
@@ -12,18 +12,13 @@ import (
 	}{
-		"new channel, existing user":     {c1, false},
-		"new channel, non-existing user": {c2, true},
+		"create new channel": {c1, false},
```
If it's only one test, just drop the map approach. However, it's a bit strange to remove the perfectly reasonable test.

anovakovic01 replied:
Currently, there are no relationships between users and channels in database. That is why this test case lost it's purpose.

mijicd commented on users/api/http/responses.go:
```diff
@@ -0,0 +1,25 @@
+
+import "net/http"
+
+type apiRes interface {
```
Since this is repeated in multiple services, I'd propose to export it in the top-level package.

mijicd commented on users/bcrypt/doc.go:
```diff
@@ -0,0 +1,2 @@
+// Package bcrypt provides a hasher implementation utilising bcrypt.
```
This can be removed since the package has only one file (i.e. move to `hasher.go`).

mijicd commented on users/jwt/doc.go:
```diff
@@ -0,0 +1,2 @@
+// Package jwt provides a JWT identity provider.
```
Same here.

mijicd commented on users/mocks/security.go:
```diff
@@ -0,0 +1,57 @@
+package mocks
```
I know it's exactly the same as in manager service, but please split `security.go` into `hasher.go` and `idp.go`.

mijicd commented on users/security.go:
```diff
@@ -0,0 +1,22 @@
+package users
```
Same remark as for `mocks/security.go`.

anovakovic01 committed:
Move service creation and config loading to separate functions

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Remove unnecessary docs files

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Split security file on hasher and idp files

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Remove test cases map from single case tests

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Move APIRes interface to project root

Update all http responses to use this interface.

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Rename transport_test to endpoint_test

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

mijicd commented on api.go:
```diff
@@ -0,0 +1,8 @@
+package mainflux
+
+// APIRes contains http response specifig methods.
+type APIRes interface {
```
Godoc missing on methods. Rename to `Response`.

anovakovic01 committed:
Rename APIRes to response and add godoc for methods

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Add gRPC error encoding to users service

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

anovakovic01 committed:
Rename identify handler and refactor gRPC server

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>
